### ISSUE TITLE
Apps/Workloads stuck in "Waiting on foregroundDeletion"

### ISSUE THREAD

**Issue by oskapt** [rancher/rancher#14394]
**Rancher versions:**

2.0.4

I had a large number of workloads that I deleted with `kubectl delete`. They all fell into a state of "Waiting on foregroundDeletion." 

I believe that this happened because I had scaled them to 0 replicas a day ago while testing a different version of the application running inside. Workloads with no replicas would not delete. If I scaled it to 1 replica and immediately deleted the workload, the deletion passed through foregroundDeletion and completed.

I was able to finalize the deletion of the stuck workloads by patching the deployment and removing the `finalizer` key (Note that `miner-1a` matched the stuck workloads):

```
kubectl -n monitoring get deploy | grep miner-1a | awk '{ print $1 }' | while read x; do kubectl -n monitoring patch deploy/$x -p '{"metadata":{"finalizers":null}}'; done
```

For the remaining workloads that still had 0 replicas I scripted the scale/delete, and they all deleted fine (Note that `miner-1c` matched the workloads that were not stuck but still had 0 replicas):

```
kubectl -n monitoring get deploy | grep miner-1c | awk '{ print $1 }' | while read x; do kubectl -n monitoring scale deploy/$x --replicas=1; sleep 1; kubectl -n monitoring delete deploy/$x; done
```

**Comment by superseb** [comment]
Stuck finalizers are a common problem, we might be able to catch these cases. Needs investigation on upstream and what we can do in Rancher

**Comment by superseb** [comment]
Possibly related: https://github.com/kubernetes/kubernetes/issues/60538 (although this is finalizer set on CRD)

Regarding catalog apps, you can apply the same patch as follows:

- Login to host running `rancher/rancher`
- Execute shell into the `rancher/rancher` container
- Patch the app, where PROJECTID and NAME can be derived from the URL of the stuck app: (`apps/p-4tkcz:efk` = PROJECTID:`p-4tkcz`, NAME=`efk`) 

```
kubectl patch apps.project.cattle.io -n PROJECTID NAME -p '{"metadata":{"finalizers":[]}}' --type=merge
```

Keep in mind that these are all unsupported options, we will need to put in an official fix for this but it needs investigation and work.

**Comment by deniseschannon** [comment]
This seems like a duplicate of https://github.com/rancher/rancher/issues/11991 (Even though one is clsuter specific vs. app/sworkloads.

**Comment by oskapt** [comment]
@alena1108 i know this is closed now, but for the record it was RKE / 1.10.1. i haven't run into it again since going through the bulk delete in my initial comment. if it comes up again, i'll add details to #11991.
